src/cpp/core/Database.cpp

```diff
@@ -0,0 +1,665 @@
+   // if this takes a long time, because we want to ensure that if we are in a hang
+   // condition (i.e. threads are not properly returning connections to the pool) we
+   // let the users/developers know that something is fishy
+   boost::shared_ptr<Connection> connection;
+   while (true)
+   {
+      if (connections_.deque(&connection, boost::posix_time::seconds(30)))
+      {
+         // create wrapper PooledConnection around retrieved Connection
+         return boost::shared_ptr<IConnection>(new PooledConnection(shared_from_this(), connection));
+      }
+      else
+      {
+         LOG_ERROR_MESSAGE("Potential hang detected: could not get database connection from pool "
+                           "after 30 seconds. If issue persists, please notify RStudio Support");
+      }
+   }
+}
+
+bool ConnectionPool::getConnection(boost::shared_ptr<IConnection>* pConnection,
+                                   const boost::posix_time::time_duration& maxWait)
+{
+   boost::shared_ptr<Connection> connection;
+   if (!connections_.deque(&connection, maxWait))
+      return false;
+
+   pConnection->reset(new PooledConnection(shared_from_this(), connection));
+   return true;
+}
+
+void ConnectionPool::returnConnection(const boost::shared_ptr<Connection>& connection)
+{
+   connections_.enque(connection);
+}
+
+Transaction::Transaction(const boost::shared_ptr<IConnection>& connection) :
+   connection_(connection),
+   transaction_(connection->session())
+{
+}
+
+void Transaction::commit()
+{
+   transaction_.commit();
+}
+
+void Transaction::rollback()
+{
+   transaction_.rollback();
+}
+
+SchemaUpdater::SchemaUpdater(const boost::shared_ptr<IConnection>& connection,
+                             const FilePath& migrationsPath) :
+   connection_(connection),
+   migrationsPath_(migrationsPath)
+{
+}
+
+Error SchemaUpdater::migrationFiles(std::vector<FilePath>* pMigrationFiles)
+{
+   std::vector<FilePath> children;
+   Error error = migrationsPath_.getChildren(children);
+   if (error)
+      return error;
+
+   for (const FilePath& file : children)
+   {
+      std::string extension = file.getExtensionLowerCase();
+      if (extension == SQL_EXTENSION ||
+          extension == SQLITE_EXTENSION ||
+          extension == POSTGRESQL_EXTENSION)
+      {
+         pMigrationFiles->push_back(file);
+      }
+   }
+
+   return Success();
+}
+
+Error SchemaUpdater::highestMigrationVersion(std::string* pVersion)
+{
+   std::vector<FilePath> files;
+   Error error = migrationFiles(&files);
+   if (error)
+      return error;
+
+   if (files.empty())
+   {
+      // no migration files - we do not consider this an error, but instead
+      // simply consider that this database cannot be migrated past version 0
+      *pVersion = "0";
+      return Success();
+   }
+
+   // sort descending - highest version filename wins
+   auto comparator = [](const FilePath& a, const FilePath& b)
+   {
+      return a.getStem() > b.getStem();
```

Comment:
Isn't this going to use character ordering ("2" > "11")?

Reply:
Upon further review I see that the filenames are formatted by convention so that character ordering is sufficient. I think we might make this clearer by using an "initial version" that looks like the incremental additional versions, and/or adding some docs stating explicitly how to format the filenames so this constraint is maintained.